core/services/gateway/handlers/workflow/handler.go

```diff
@@ -0,0 +1,152 @@
+package workflow
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/multierr"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/api"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/config"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/handlers"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/network"
+)
+
+const (
+	MethodAddWorkflow = "add_workflow"
+	// NOTE: more methods will go here: CRUD for workflow specs; HTTP trigger/action/target; etc.
+	MethodWebAPITarget = "web_api_target"
+)
+
+type workflowHandler struct {
+	donConfig      *config.DONConfig
+	don            handlers.DON
+	savedCallbacks map[string]*savedCallback
+	mu             sync.Mutex
+	lggr           logger.Logger
+	httpClient     network.HttpClient // for outgoing requests to users
+}
+
+type savedCallback struct {
+	id         string
+	callbackCh chan<- handlers.UserCallbackPayload
+}
+
+var _ handlers.Handler = (*workflowHandler)(nil)
+
+func NewWorkflowHandler(donConfig *config.DONConfig, don handlers.DON, httpClient network.HttpClient, lggr logger.Logger) (*workflowHandler, error) {
+	return &workflowHandler{
+		donConfig:      donConfig,
+		don:            don,
+		savedCallbacks: make(map[string]*savedCallback),
+		httpClient:     httpClient,
+		lggr:           lggr.Named("WorkflowHandler." + donConfig.DonId),
+	}, nil
+}
+
+func (d *workflowHandler) HandleUserMessage(ctx context.Context, msg *api.Message, callbackCh chan<- handlers.UserCallbackPayload) error {
+	d.mu.Lock()
+	d.savedCallbacks[msg.Body.MessageId] = &savedCallback{msg.Body.MessageId, callbackCh}
+	don := d.don
+	d.mu.Unlock()
+
+	// TODO: apply allowlist and rate-limiting here.
+	if msg.Body.Method != MethodAddWorkflow {
+		d.lggr.Errorw("unsupported method", "method", msg.Body.Method)
+		return fmt.Errorf("unsupported method")
+	}
+
+	var err error
+	// Send to all nodes.
+	for _, member := range d.donConfig.Members {
+		err = multierr.Combine(err, don.SendToNode(ctx, member.Address, msg))
+	}
+	return err
+}
+
+func (d *workflowHandler) handleWebAPITargetMessage(ctx context.Context, msg *api.Message, nodeAddr string) error {
+	var targetPayload TargetRequestPayload
+	err := json.Unmarshal(msg.Body.Payload, &targetPayload)
+	if err != nil {
+		return err
+	}
+	// send message to target
+	req := network.HttpRequest{
+		Method:     targetPayload.Method,
+		URL:        targetPayload.URL,
+		Headers:    targetPayload.Headers,
+		Body:       targetPayload.Body,
+		Timeout:    time.Duration(targetPayload.TimeoutMs) * time.Millisecond,
+		RetryCount: targetPayload.RetryCount,
```

Comment:
I'm not sure how this is generated, but this should be different than the retrycount from gateway.  And it doesnt make sense to retry on many 4xx codes like 401, 403, 404

Reply:
This will be the customer-provided retries. I agree that there are 2 level of retries needed:
1. user-defined retries - this is defined in the workflows
2. node-level retries - retries between gateway and node.

decided to leave a TODO for retry logic but yes, fully agreed we shouldn't retry on all errors